libvast/src/chunk.cpp

```diff
@@ -54,16 +53,20 @@ chunk_ptr chunk::make(view_type view, deleter_type&& deleter) noexcept {
   return chunk_ptr{new chunk{view, std::move(deleter)}, false};
 }
 
-chunk_ptr chunk::mmap(const path& filename, size_type size, size_type offset) {
+chunk_ptr chunk::mmap(const std::filesystem::path& filename, size_type size,
+                      size_type offset) {
   // Figure out the file size if not provided.
   if (size == 0) {
-    auto s = file_size(filename);
-    if (!s)
+    std::error_code err{};
+    size = std::filesystem::file_size(filename, err);
+    if (size == static_cast<std::uintmax_t>(-1)) {
+      VAST_INFO("failed to get file size for filename {}: {}", filename,
+                err.message());
```

Comment:
```suggestion
      VAST_ERROR("failed to get file size for filename {}: {}", filename,
                 err.message());
```

Reply:
This is now the only blocker for this PR, I think.

I wonder whether it makes sense to change the return value of `chunk::mmap` to be `caf::expected<chunk_ptr>`, with the post-condition that if the return value is not an error, it is never nullptr. What do you think @mavam?

Reply:
The general guideline should be: if a binary distinction between `nullptr` and correctly created object is not enough information to understand failed construction at the call site, then going through the `expected` monad is reasonable to enrich the error. In other words, if there a multiple causes failure and the call site wants to distinguish them, then `expected` makes sense.

Reply:
I disagree. I personally prefer returning an error here because that can be forwarded to the client process so the user actually sees it.

Reply:
My point is different: if you know exactly what failed, you can also generate the error at the call site. There is no value in `expected` in this case. If not, only then `expected` adds value.

Reply:
Then printing here at all leads to duplicate error reporting.

Reply:
Sorry, but now I disagree 🙂 . Let's apply the rationale by looking at the implementation:, which has 3 calls that can fail:

1. `file_size`
2. `open`
3. `mmap`

The call site doesn't know what failed when it receives `nullptr`. So we are in the second scenario that I described. Ergo: I advocate that we *do* need `expected<chunk_ptr>` here.

Reply:
(As a result, there should not be logging in here because it will likely be done at the call site.)

Reply:
For now, I've changed it to `VAST_ERROR`. I think if we want to change the return value of this API, let's do that in a small, follow-up PR so it can be reviewed in isolation.

In general, should all callers of `std::filesystem::file_size` report an error if we can't get the file size?

FWIW, I don't think callers should be generating the error at the call site as that requires knowledge of the function implementation (e.g. did the file size check fail due to permissions, file not existing, etc...). If we want to go the `expected` route, it should carry the appropriate `err.message()` so the caller can consume it appropriately.

Reply:
> For now, I've changed it to `VAST_ERROR`. I think if we want to change the return value of this API, let's do that in a small, follow-up PR so it can be reviewed in isolation.

Agreed.

> In general, should all callers of `std::filesystem::file_size` report an error if we can't get the file size?

Depends on the call site: some times it's a dealbreaker but sometimes it doesn't hurt. In practice we probably want to report this to `ERROR` or `WARNING` log levels.
 
> FWIW, I don't think callers should be generating the error at the call site as that requires knowledge of the function implementation (e.g. did the file size check fail due to permissions, file not existing, etc...). If we want to go the `expected` route, it should carry the appropriate `err.message()` so the caller can consume it appropriately.

I think we're saying the same thing here. The error handling at the call site is uniform, but you still want to know what *exactly* failed if there are multiple options. By consuming the error message, users get actionable insight in what went wrong.

Reply:
Yep, we're on the same page @mavam. I'll work on changing this API to return `expected` and introduce a follow-up PR on top of this soon.

Are we all good with this PR for now?

Reply:
All good afaict. I'll let @mavam merge.

Reply:
See https://github.com/tenzir/vast/pull/1509 for the follow-up PR. I'll rebase it once this PR lands.